### ISSUE TITLE
Add preprocessor layer for neural network

### ISSUE THREAD

**Issue by lisitsyn** [shogun-toolbox/shogun#2710]
We now have all the basic layers like RLU, sigmoid and various output layers but no way to apply some intermediate preprocessing like PCA. This could be useful sometimes so this task suggests to add a layer that applies some CPreprocessor on data (feed-forward step). This layer should have no effect on neural net weights so it should pass gradients through on back-propagation step.

Good for [Deep learning](https://github.com/shogun-toolbox/shogun/wiki/GSoC_2015_project_deep_learning) applicants

**Comment by arasuarun** [comment]
Do we have a completely new Network class which has a Preprocessor (say DimensionReduction) member (sent via enum or object) and applies that on the given input, generating an output which is now the input for the rest of the network? This preprocessing has no part to play in the remainder of the training, right?

**Comment by lisitsyn** [comment]
@arasuarun yes it is would be like a non-learnable layer, that transforms data though

**Comment by arasuarun** [comment]
Should we add Gaussian noise to the data before or after preprocessing? (referring to **compute_activations** in [Neural Input Layer](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/neuralnets/NeuralInputLayer.cpp))

My idea is to derive *Neural Preprocessor Layer* from *Neural Input Layer* and do the preprocessing step in the compute_activations method. This would be suitable for *DimensionReductionPreprocesor*s. The user would pass in the object of the exact subclass.

Is this method fine? What other kinds of preprocessing are you looking at?

**Comment by lisitsyn** [comment]
@arasuarun I think gaussian noise should be made up of just another layer, could be done later.

Anyway, wouldn't making it input layer restrict us from putting it somewhere between other layers?

**Comment by arasuarun** [comment]
@lisitsyn (Please excuse the late response) I'm a bit confused in the implementation. I think it'd be better if I had a better view of **why** this would be used. Could you give me any links to check out? (cause I couldn't find anything related myself)
